Terminal.Gui/Views/TextField.cs

```diff
@@ -637,36 +631,44 @@ void SetClipboard (IEnumerable<Rune> text)
 		///    </item>
 		/// </list>
 		/// </remarks>
-		public override bool ProcessKey (KeyEvent kb)
+		/// </summary>
+		/// <param name="a"></param>
+		/// <returns></returns>
+		public override bool OnKeyPressed (KeyEventArgs a)
 		{
 			// remember current cursor position
 			// because the new calculated cursor position is needed to be set BEFORE the change event is triggest
 			// Needed for the Elmish Wrapper issue https://github.com/DieselMeister/Terminal.Gui.Elmish/issues/2
 			_oldCursorPos = _point;
 
 			// Give autocomplete first opportunity to respond to key presses
-			if (SelectedLength == 0 && Autocomplete.Suggestions.Count > 0 && Autocomplete.ProcessKey (kb)) {
+			if (SelectedLength == 0 && Autocomplete.Suggestions.Count > 0 && Autocomplete.ProcessKey (a)) {
+				return true;
+			}
+
+			if (base.OnKeyPressed (a)) {
 				return true;
 			}
 
-			var result = InvokeKeybindings (new KeyEvent (ShortcutHelper.GetModifiersKey (kb),
-				new KeyModifiers () { Alt = kb.IsAlt, Ctrl = kb.IsCtrl, Shift = kb.IsShift }));
-			if (result != null)
-				return (bool)result;
+			if (OnInvokeKeyBindings (a)) {
+				return true;
+			}
```

Comment:
Got it. Currently:

InvokeKeyBindings:
![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/f6bba5a5-c41a-4d61-8039-aacf1208eb69)

In the current impl of `OnKeyBindings`:
![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/27a5d02a-9df0-42b6-a84b-356ae4585783)

If no keybinding was found, continue.
If keybindings were found, but  none handled the key, continue.
If keybindings were found, and at least one handled the key, stop.

New logic:

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/2fd59e01-2edf-4932-ba34-a42104191eb8)

![image](https://github.com/gui-cs/Terminal.Gui/assets/585482/c60d8a32-90ff-400d-b4c0-afa44ff58123)

Correct?

Reply:
No, if it's false we may want to return false without resuming the rest. I know this is complex and is one of the things which is giving me more work.
In your sample you should return (bool)handled if not null, handled itself otherwise.

Reply:
Why? What's the user scenario where a key binding was found, invoked, but not handled and processing should stop?

Reply:
I'm referring only the returning value when it's called.

Reply:
> Why? What's the user scenario where a key binding was found, invoked, but not handled and processing should stop?

In the virtual method all the key bindings that were found will return `true` or `false`, otherwise return `null`, so they were handled and the processing should stop. If a view want to do some custom action after the must override it.

Reply:
I'm still not getting the point you're trying to make.

* If no keybinding was found, `InvokeKeyBindings` returns `null`. Continue passing the event (return `false` from `OnInvokeKeyBindings`).
* If keybindings were found, but none handled the key (all `Command`s returned `false`), `InvokeKeyBindings` returns `false`. Continue passing the event (return `false` from `OnInvokeKeyBindings`)..
* If keybindings were found, and any handled the key (at least one `Command` returned `true`),
`InvokeKeyBindings` returns `true`. Continue passing the event (return `false` from `OnInvokeKeyBindings`).

Here's what I'm proposing that, I believe, does exactly what these rules say:

```cs
public bool ProcessKeyPressEvent (KeyEventArgs keyEvent)
{
	if (!Enabled) {
		return false;
	}

	if (Focused?.ProcessKeyPressEvent (keyEvent) == true) {
		return true;
	}

	// Before (fire the cancellable event KeyPress)
	if (OnKeyPress (keyEvent)) {
		return true;
	}

	// During (this is what can be cancelled)
	// Overridable method that can be used to process a key press.
	// (fires the cancellable event ProcessKeyPress)
	var handled = OnInvokeKeyBindings (keyEvent);
	if (handled != null && (bool)handled) {
		return true;
	}
	
	if (OnProcessKeyPress (keyEvent)) {
		return true;
	}

	// After (fire the cancellable event InvokingKeyBindings)


	return false;
}

...

public virtual bool? OnInvokeKeyBindings (KeyEventArgs keyEvent)
{
	// fire event
	InvokingKeyBindings?.Invoke (this, keyEvent);
	if (keyEvent.Handled) {
		return true;
	}

	var handled = InvokeKeyBindings (keyEvent);
	if (handled != null && (bool)handled) {
		return true;
	}
	return false;
}
```

You may have misread the code images I pasted above, and the actual text here is more clear.

Reply:
This method must also return `null` if it's the case. So, in the `ProcessKeyPressEvent` if handled is true return true and if it's false or null return false. The reason for this is if a view that call `OnInvokeKeyBindings` can know if a key binding was found (true or false) or not found at all and so it's null.

```cs
public virtual bool? OnInvokeKeyBindings (KeyEventArgs keyEvent)
{
	// fire event
	InvokingKeyBindings?.Invoke (this, keyEvent);
	if (keyEvent.Handled) {
		return true;
	}

	var handled = InvokeKeyBindings (keyEvent);
	if (handled != null) {
		return (bool)handled;
	}
	return handled;
}
```

Reply:
Duh. Thanks 🙏!